solarforecastarbiter/datamodel.py

```diff
@@ -1017,21 +1017,25 @@ def __check_metrics__(fx, metrics):
 
     Raises
     ------
-    TypeError
+    ValueError
         If the selected metrics are not valid for the given forecast type.
 
     TODO
     ----
-    * validate probabilistic forecast metrics
     * validate event forecast metrics
 
     """
 
-    if isinstance(fx, Forecast) and not set(metrics) <= \
-            ALLOWED_DETERMINISTIC_METRICS.keys():
-        raise ValueError("Metrics must be in ALLOWED_DETERMINISTIC_METRICS")
-    else:
-        pass
+    ALLOWED_DETERMINISTIC_FORECASTS = [Forecast]
+    ALLOWED_PROBABILISTIC_FORECASTS = [ProbabilisticForecast,
+                                       ProbabilisticForecastConstantValue]
+
+    if type(fx) in ALLOWED_DETERMINISTIC_FORECASTS and not set(metrics) \
```

Comment:
I think `isinstance` is preferred to `type` in `if` statements for reasons around inheritance. Is there a reason you've changed to `type`?

Reply:
Along with that, I think its more clear just to put the class in there instead of making ``ALLOWED_..._FORECASTS``

Reply:
> I think `isinstance` is preferred to `type` in `if` statements for reasons around inheritance. Is there a reason you've changed to `type`?

While developing the code, I was running into issues where a `isinstance` check on a probabilistic `fx` would mark the `fx` as deterministic since `ProbabilisticForecast` inherits from `Forecast`. But let me double-check whether that is still the case. (The root issue I was seeing may have been a side-effect of another issue, which I may have now fixed.)

Reply:
> Along with that, I think its more clear just to put the class in there instead of making `ALLOWED_..._FORECASTS`

I can explicitly list out the forecast classes, if that's preferred. I originally switched to the variable version because I thought that was clearer given it required less line breaks to enforce the max 80-character limit. Let me switch to the more explicit list version and then we can review.

Reply:
**Update**: I tried switching to `isinstance` and I'm running into the same issue of `isinstance` marking `fx` of type `ProbabilisticForecast` or `ProbabilisticForecastConstantValue` as `Forecast` (since both inherit from `Forecast`). Perhaps there is a way to modify how we call `isinstance` to avoid this issue, but I'm not aware of an easy fix besides just using `if fx in <list of types>`.

Reply:
Regarding the `ALLOWED_..._FORECASTS` variables: here's what it looks like with listing out the forecast types in-line:

https://github.com/SolarArbiter/solarforecastarbiter-core/blob/2c5139333feb82134e5be0f18aee8d5186c99a4c/solarforecastarbiter/datamodel.py#L1029-L1035

I personally think the `ALLOWED_..._FORECASTS` version is cleaner and easier to understand/read.

Reply:
For direct comparison, here's what the two versions look like:

```python
    if type(fx) in [Forecast] and not set(metrics) <= \
            ALLOWED_DETERMINISTIC_METRICS.keys():
        raise ValueError("Metrics must be in ALLOWED_DETERMINISTIC_METRICS.")
    elif type(fx) in [ProbabilisticForecast, \
            ProbabilisticForecastConstantValue] and not set(metrics) <= \
            ALLOWED_PROBABILISTIC_METRICS.keys():
        raise ValueError("Metrics must be in ALLOWED_PROBABILISTIC_METRICS.")
```


```python
    ALLOWED_DETERMINISTIC_FORECASTS = [Forecast]
    ALLOWED_PROBABILISTIC_FORECASTS = [ProbabilisticForecast,
                                       ProbabilisticForecastConstantValue]

    if type(fx) in ALLOWED_DETERMINISTIC_FORECASTS and not set(metrics) <= \
            ALLOWED_DETERMINISTIC_METRICS.keys():
        raise ValueError("Metrics must be in ALLOWED_DETERMINISTIC_METRICS.")
    elif type(fx) in ALLOWED_PROBABILISTIC_FORECASTS and not set(metrics) <= \
            ALLOWED_PROBABILISTIC_METRICS.keys():
        raise ValueError("Metrics must be in ALLOWED_PROBABILISTIC_METRICS.")
```

Reply:
Also, we could use different names for `ALLOWED_..._FORECASTS` (e.g. rename `ALLOWED_DETERMINISTIC_FORECASTS` as `DETERMINISTIC_FORECAST_TYPES`) and the use of the two `ALLOWED_..._FORECASTS` variables makes it easier (in my opinion) to extend in the future to cover other forecast dataclasses.

Reply:
Can you just check if ``isinstance`` for the probabilistic ones first? I don't care so much about the lists but the variable names should be lowercase. All caps usually signals a global variable. Also btw, you can is parentheses to split a multiline if conditional and put a tuple of types in isinstance

Reply:
I tried changing the `isinstance` order before and that doesn't work (not 100% sure why, since I agree it should work in theory). I'll push a commit with the `isinstance` order changed to confirm it's not just something weird on my local machine.

Reply:
The problem then is likely that you need to split the if conditional i.e. first test if isinstance then another if for the metric set. Otherwise it probably bypasses the first case but is still a forecast instance and the metrics are wrong.

I think we definitely want to use isinstance here over type

Reply:
I disagree with this solution, but I'll defer to your/Will's judgement as leads on the project and will push a commit with the required revisions.

Reply:
With splitting the if or using isinstance over type? Or both?  I don't think there is one right answer, just curious about your reasoning 😃